piecrust/src/store.rs

```diff
@@ -0,0 +1,903 @@
+/// current state of the memory. These hashes then form the leaves of the tree
+/// whose root is then computed.
+fn compute_root<'a, I>(
+    base: &Option<(Root, Commit)>,
+    modules: I,
+) -> (Root, BTreeMap<ModuleId, Root>)
+where
+    I: IntoIterator<Item = (&'a ModuleId, &'a (Bytecode, Memory))>,
+{
+    let iter = modules.into_iter();
+
+    let mut leaves_map = BTreeMap::new();
+
+    // Compute the hashes of changed memories
+    for (module, (_, memory)) in iter {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&memory.read());
+        leaves_map.insert(*module, Root::from(hasher.finalize()));
+    }
+
+    // Store the hashes of *un*changed memories
+    if let Some((_, base_commit)) = base {
+        for (module, root) in &base_commit.modules {
+            if !leaves_map.contains_key(module) {
+                leaves_map.insert(*module, *root);
+            }
+        }
+    }
+
+    let mut leaves: Vec<Root> = leaves_map.clone().into_values().collect();
+
+    while leaves.len() > 1 {
+        leaves = leaves
+            .chunks(2)
+            .map(|chunk| {
+                let mut hasher = blake3::Hasher::new();
+
+                hasher.update(&chunk[0]);
+                if chunk.len() > 1 {
+                    hasher.update(&chunk[1]);
+                }
+
+                Root::from(hasher.finalize())
+            })
+            .collect();
+    }
+
+    (leaves[0], leaves_map)
+}
+
+/// Delete the given commit's directory.
+fn delete_commit_dir<P: AsRef<Path>>(
+    root_dir: P,
+    root: Root,
+) -> io::Result<()> {
+    let root = hex::encode(root);
+    let commit_dir = root_dir.as_ref().join(root);
+    fs::remove_dir_all(commit_dir)
+}
+
+/// Squash the given commit.
+fn squash_commit<P: AsRef<Path>>(
+    root_dir: P,
+    root: Root,
+    commit: &mut Commit,
+) -> io::Result<()> {
+    let root_dir = root_dir.as_ref();
+
+    let root_hex = hex::encode(root);
+
+    let commit_dir = root_dir.join(root_hex);
+
+    let memory_dir = commit_dir.join(MEMORY_DIR);
+
+    for module in &commit.diffs {
+        let module_hex = hex::encode(module);
+        let memory_path = memory_dir.join(module_hex);
+        let memory_diff_path = memory_path.with_extension(DIFF_EXTENSION);
+
+        let memory =
+            Memory::from_file_and_diff(&memory_path, &memory_diff_path)?;
+
+        fs::remove_file(&memory_path)?;
+        fs::remove_file(memory_diff_path)?;
+
+        fs::write(memory_path, memory.read())?;
+    }
+
+    Ok(())
+}
+
+/// The representation of a session with a [`ModuleStore`].
+///
+/// A session tracks modifications to the modules' memories by keeping
+/// references to the set of instantiated modules.
+///
+/// The modifications are kept in memory and are only persisted to disk on a
+/// call to [`commit`].
+///
+/// [`commit`]: ModuleSession::commit
+pub struct ModuleSession {
+    modules: BTreeMap<ModuleId, (Bytecode, Memory)>,
```

Comment:
I'd consider moving ModuleSession to a separate file, as the class is important in its own right and also the file store.rs is long already.